mobile/library/java/io/envoyproxy/envoymobile/engine/AndroidNetworkMonitor.java

```diff
@@ -70,20 +80,46 @@ public void onAvailable(@NonNull Network network) {
     @Override
     public void onCapabilitiesChanged(@NonNull Network network,
                                       @NonNull NetworkCapabilities networkCapabilities) {
-      if (networkCapabilities.hasCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET)) {
-        if (networkCapabilities.hasCapability(NetworkCapabilities.TRANSPORT_WIFI)) {
-          envoyEngine.onDefaultNetworkChanged(EnvoyNetworkType.WLAN);
-        } else if (networkCapabilities.hasCapability(NetworkCapabilities.TRANSPORT_CELLULAR)) {
-          envoyEngine.onDefaultNetworkChanged(EnvoyNetworkType.WWAN);
-        } else {
-          envoyEngine.onDefaultNetworkChanged(EnvoyNetworkType.GENERIC);
+      // `onCapabilities` is guaranteed to be called immediately after `onAvailable`.
+      // https://developer.android.com/reference/android/net/ConnectivityManager.NetworkCallback#onCapabilitiesChanged(android.net.Network,%20android.net.NetworkCapabilities)
+      if (transportType == EMPTY_TRANSPORT_TYPE) {
+        // The network was lost previously, see `onLost`.
+        onDefaultNetworkTypeChanged(networkCapabilities);
+      } else {
+        // Only call the `onDefaultNetworkTypeChanged` callback when there is a change in the
+        // transport type.
+        if (!networkCapabilities.hasTransport(transportType)) {
+          onDefaultNetworkTypeChanged(networkCapabilities);
+          transportType = getTransportType(networkCapabilities);
         }
       }
     }
 
     @Override
     public void onLost(@NonNull Network network) {
       envoyEngine.onDefaultNetworkUnavailable();
+      transportType = EMPTY_TRANSPORT_TYPE;
+    }
+
+    private static int getTransportType(NetworkCapabilities networkCapabilities) {
+      for (int type : TRANSPORT_TYPES) {
+        if (networkCapabilities.hasTransport(type)) {
+          return type;
+        }
+      }
+      return EMPTY_TRANSPORT_TYPE;
+    }
+
+    private void onDefaultNetworkTypeChanged(NetworkCapabilities networkCapabilities) {
+      if (networkCapabilities.hasCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET)) {
+        if (networkCapabilities.hasTransport(NetworkCapabilities.TRANSPORT_WIFI)) {
```

Comment:
Should TRANSPORT_WIFI_AWARE also be regarded as WLAN https://source.chromium.org/chromium/chromium/src/+/main:net/android/java/src/org/chromium/net/NetworkChangeNotifierAutoDetect.java;l=681;drc=8e948282d37c0e119e3102236878d6f4d5052c16;bpv=1;bpt=1?

Reply:
Good point. Done.